### PR TITLE
fix(upload): beforeUpload type definition

### LINKED ISSUE #30347: beforeUpload return type isn't allowing LIST_IGNORE in promise
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### Reproduction link
[Edit on CodeSandbox](https://codesandbox.io/s/antds-beforeupload-prop-type-issue-1cyrz)

### Steps to reproduce
1. Import and use the Upload component
2. Pass a beforeUpload which returns the promise of `Upload.LIST_IGNORE`
3. See TypeScript complain the types are incompatible

### What is expected?
Type annotation of the beforeUpload prop should cover all of its usage.

### What is actually happening?
The type annotation of beforeUpload incompatible with one of its usage.

| Environment | Info |
|---|---|
| antd | 4.15.3 |
| React | n/a |
| System | n/a |
| Browser | n/a |

---
The implementation of beforeUpload does support returning the promise of `Upload.LIST_IGNORE` only the type annotation is incorrect.

### PR DESCRIPTION
**Pull request by kerm1it**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [x] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

fix https://github.com/ant-design/ant-design/issues/30347

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Fix Upload type definition of `beforeUpload`.          |
| 🇨🇳 Chinese | 修复上传组件 `beforeUpload` 类型声明。          |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/upload/__tests__/type.test.tsx [hunk 1]
```diff
@@ -32,13 +32,28 @@ describe('Upload.typescript', () => {
...
-          return Upload.LIST_IGNORE;
+          if (returnType === 'string') {
+            return Upload.LIST_IGNORE;
+          }
+          if (returnType === 'Promise<string>') {
+            return Promise.resolve(Upload.LIST_IGNORE);
+          }
+          if (returnType === 'Promise<void>') {
+            return Promise.resolve();
+          }
```
Comment (afc163) [inline components/upload/__tests__/type.test.tsx:56]: 加个 await 的用例？

Comment (kerm1it) [reply components/upload/__tests__/type.test.tsx:56]: 什么意思？Promise 和 直接返回的 case 我都写了

Comment (afc163) [reply components/upload/__tests__/type.test.tsx:56]: 虽然结果一样，await 的写法难保会有 TS 的兼容性问题，还是也作为用例比较靠谱。

Comment (kerm1it) [reply components/upload/__tests__/type.test.tsx:66]: 明白了，我拆开了，一个是同步的，一个是 async 异步的

File: components/upload/__tests__/type.test.tsx [hunk 2]
```diff
@@ -36,23 +36,33 @@ describe('Upload.typescript', () => {
...
             return file;
           }
-          if (returnType === 'Promise<file>') {
-            return Promise.resolve(file);
-          }
           if (returnType === 'string') {
             return Upload.LIST_IGNORE;
           }
-          if (returnType === 'Promise<string>') {
-            return Promise.resolve(Upload.LIST_IGNORE);
```
Comment (afc163) [inline components/upload/__tests__/type.test.tsx]: 可以留着。

**Pull request merged**